pandas/core/dtypes/cast.py

```diff
@@ -264,27 +264,22 @@ def maybe_promote(dtype, fill_value=np.nan):
 
     # returns tuple of (dtype, fill_value)
     if issubclass(dtype.type, (np.datetime64, np.timedelta64)):
-        # for now: refuse to upcast datetime64
-        # (this is because datetime64 will not implicitly upconvert
-        #  to object correctly as of numpy 1.6.1)
         if isna(fill_value):
             fill_value = iNaT
+        elif issubclass(dtype.type, np.datetime64):
+            try:
+                fill_value = tslibs.Timestamp(fill_value).value
+            except Exception:
+                dtype = np.object_
+                fill_value = np.nan
+        elif issubclass(dtype.type, np.timedelta64):
+            try:
+                fill_value = tslibs.Timedelta(fill_value).value
+            except Exception:
+                dtype = np.object_
```

Comment:
are these exceptions actually hit in any tests?

Reply:
The thing is that `maybe_promote` is not tested directly (see also #23833), I'm working on a PR for that currently.

Also, I've had a segfault in `pandas/tests/test_take.py::TestTake::test_2d_datetime64` on an earlier attempt, but let's see.